cpp/src/groupby/sort/scan.cpp

```diff
@@ -115,42 +121,87 @@ template <>
 void scan_result_functor::operator()<aggregation::RANK>(aggregation const& agg)
 {
   if (cache.has_result(values, agg)) return;
-  CUDF_EXPECTS(helper.is_presorted(),
-               "Rank aggregate in groupby scan requires the keys to be presorted");
-  auto const order_by = get_grouped_values();
-  CUDF_EXPECTS(!cudf::structs::detail::is_or_has_nested_lists(order_by),
-               "Unsupported list type in grouped rank scan.");
+  // TODO review: this denotes key sorted, but the values are not sorted.
+  // CUDF_EXPECTS(helper.is_presorted(),
+  //              "Rank aggregate in groupby scan requires the keys to be presorted");
 
-  cache.add_result(
-    values,
-    agg,
-    detail::rank_scan(
-      order_by, helper.group_labels(stream), helper.group_offsets(stream), stream, mr));
+  CUDF_EXPECTS(!cudf::structs::detail::is_or_has_nested_lists(values),
+               "Unsupported list type in grouped rank scan.");
+  auto const& rank_agg         = dynamic_cast<cudf::detail::rank_aggregation const&>(agg);
+  auto const& group_labels     = helper.group_labels(stream);
+  auto const group_labels_view = column_view(cudf::device_span<const size_type>(group_labels));
+  auto const gather_map =
+    is_presorted()
+      ? cudf::detail::sequence(
+          group_labels.size(), *cudf::make_fixed_width_scalar(size_type{0}, stream), stream)
+      : ((rank_agg._method == rank_method::FIRST
+            ? cudf::detail::stable_sorted_order
+            : cudf::detail::sorted_order)(table_view({group_labels_view, get_grouped_values()}),
+                                          {order::ASCENDING, rank_agg._column_order},
+                                          {null_order::AFTER, rank_agg._null_precedence},
+                                          stream,
+                                          rmm::mr::get_current_device_resource()));
+
+  auto rank_scan = [&]() {
+    if (rank_agg._method == rank_method::MIN) {
```

Comment:
Would this be less verbose if it were a `switch case`?

Reply:
That's right!  I kept adding more methods and didn't think to replace with `switch`.